File: 문제풀이/온라인 저지/프로그래머스/# 01. Level 1/# 064061. 크레인 인형뽑기 게임/Solution.java
import java.util.Stack;

public class Solution {

    private static int answer = 0;
    public int solution(int[][] board, int[] moves) {
        Stack<Integer> stack = new Stack<>();
        int position;
        int doll;

        for (int move : moves) {
            position = calculatePositionIndex(move);
            doll = extractDoll(board, position);
            putDoll(stack, doll);
        }
        return answer;
    }

    private static int calculatePositionIndex(int move) {
        return move - 1;
    }

    private static int extractDoll(int[][] board, int position) {
        int current = 0;
        for (int i=0; i<board.length; i++) {
            current = board[i][position];
            if (current > 0) {
                board[i][position] = 0;
                return current;
            }
        }
        return current;
    }

    private void putDoll(Stack<Integer> stack, int doll) {
        if (doll == 0) {
            return;
        }
        if (!stack.isEmpty() && stack.peek() == doll) {
            stack.pop();
            answer += 2;
            return;
        }
        stack.push(doll);
    }
}

File: 문제풀이/온라인 저지/프로그래머스/# 01. Level 1/# 064061. 크레인 인형뽑기 게임/python/solution.py
from typing import List

count = 0


def solution(board: List[List[int]], moves: int) -> int:
    stack = []

    for move in moves:
        index = move - 1
        doll = extract_doll(board, index)
        put_doll(stack, doll)

    return count


def extract_doll(board: List[List[int]], index: int) -> int:
    current = 0
    for i in range(0, len(board)):
        current = board[i][index]
        if current > 0:
            board[i][index] = 0
            return current

    return current


def put_doll(stack: List[List[int]], doll: int) -> None:
    global count

    if doll == 0:
        return

    if len(stack) != 0 and stack[-1] == doll:
        stack.pop()
        count += 2
        return

    stack.append(doll)
